<a href="https://colab.research.google.com/github/jarvispraba/datakelompok/blob/main/KKA_XI_RPL_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
class Kucing:
    # Constructor
    def __init__(self, nama, warna):
        self.nama = nama # Attribute
        self.warna = warna # Attribute
    # Method
    def suara(self):
        return f"{self.nama} berkata: Meow!"
# Membuat Object (Instansiasi)
kucing1 = Kucing("Luna", "Putih")
kucing2 = Kucing("Garfield", "Oranye")
# Mengakses Attribute dan Method
print(kucing1.nama) # Output: Luna
print(kucing2.suara()) # Output: Garfield berkata: Meow!

Luna
Garfield berkata: Meow!


In [ ]:
# Parent Class
class Hewan:
  def makan(self):
    print("Hewan ini sedang makan.")

# Child Class (Mewarisi Hewan)
class Burung(Hewan):
  def terbang(self):
    print("Burung sedang terbang.")
# Object
beo = Burung()
beo.makan() # Mewarisi dari Hewan (Output: Hewan ini sedang makan.)
beo.terbang() # Method milik sendiri

Hewan ini sedang makan.
Burung sedang terbang.


In [ ]:
class AkunBank:
  def __init__(self, saldo):
    self.__saldo = saldo # Private attribute

  def cek_saldo(self):
    return self.__saldo

  def setor(self, jumlah):
    if jumlah > 0:
      self.__saldo += jumlah

rekening = AkunBank(1000)
rekening.setor(500)
print(rekening.cek_saldo()) # Output: 1500
# print(rekening.__saldo) # Ini akan Error! Tidak bisa akses langsung.

1500


In [ ]:
class Anjing:
  def suara(self):
    return "Guk Guk!"
class Kucing:
  def suara(self):
    return "Meow!"
# Fungsi umum
def tes_suara(hewan):
  print(hewan.suara())
hewan1 = Anjing()
hewan2 = Kucing()
tes_suara(hewan1) # Output: Guk Guk!
tes_suara(hewan2) # Output: Meow!

Guk Guk!
Meow!


In [ ]:
from abc import ABC, abstractmethod
class Kendaraan(ABC): # Class Abstrak
  @abstractmethod
  def bergerak(self):
    pass
class Mobil(Kendaraan):
  def bergerak(self):
    print("Berjalan dengan roda")
class Kapal(Kendaraan):
  def bergerak(self):
    print("Berlayar di air")

# k = Kendaraan() # Ini akan Error! Class abstrak tidak bisa di instansiasi.
m = Mobil()
m.bergerak()

Berjalan dengan roda


In [ ]:
from abc import ABC, abstractmethod
# Ini bertindak sebagai Interface
class TombolInterface(ABC):
  @abstractmethod
  def tekan(self):
    pass
# Class ini WAJIB punya method tekan()
class TombolMerah(TombolInterface):
  def tekan(self):
    print("Sistem Dimatikan!")
class TombolHijau(TombolInterface):
  def tekan(self):
    print("Sistem Dinyalakan!")

--------------------------------------------------------------

**LATIHAN SOAL 1-2**

In [1]:
class Hero:
# Constructor: Dijalankan saat Hero baru dibuat
  def __init__(self, name, hp, attack_power):
    self.name = name # Nama Hero
    self.hp = hp # Nyawa (Health Point)
    self.attack_power = attack_power # Kekuatan Serangan
  # Method untuk menampilkan info hero
  def info(self):
    print(f"Hero: {self.name} | HP: {self.hp} | Power: {self.attack_power}")

  # Method menyerang: Objek ini (self) menyerang objek lain (lawan)
  def serang(self, lawan):
      print(f"{self.name} menyerang {lawan.name}!")
      lawan.diserang(self.attack_power)

  # Method diserang: Menerima damage
  def diserang(self, damage):
      self.hp -= damage
      print(f"{self.name} terkena damage {damage}. Sisa HP: {self.hp}")

# -- Main Program --
# Membuat Object (Instansiasi)
hero1 = Hero("Layla", 100, 15)
hero1.hp = 500
hero2 = Hero("Zilong", 120, 20)
# Memanggil Method
hero1.info()
hero2.info()

# Output di akhir program
print("\n--- Pertarungan Dimulai ---")
hero1.serang(hero2)  # Layla menyerang Zilong
hero2.serang(hero1)  # Zilong membalas

# Display final HP (added for completeness)
print("\n--- Statistik Akhir ---")
hero1.info()
hero2.info()

Hero: Layla | HP: 500 | Power: 15
Hero: Zilong | HP: 120 | Power: 20

--- Pertarungan Dimulai ---
Layla menyerang Zilong!
Zilong terkena damage 15. Sisa HP: 105
Zilong menyerang Layla!
Layla terkena damage 20. Sisa HP: 480

--- Statistik Akhir ---
Hero: Layla | HP: 480 | Power: 15
Hero: Zilong | HP: 105 | Power: 20


Hero Layla HP berubah menjadi 500, dari yang awalnya hanya 100.

Karena lawan menerima objek utuh, method serang dapat mengakses atribut dan method milik lawan, seperti lawan.name dan lawan.diserang. Dengan begitu, serangan bisa langsung mengurangi HP objek lawan.

Dan dari yg awalnya hp layla 500 menjadi 480 karena terkena damage, sama halnya dengan hp zilong dari 120 menjadi 105. Karena HP dari hero tersebut berkurang oleh def diserang dan attackpower dari hero yg menyerang

### **LATIHAN SOAL 3**

In [3]:
# Class Mage adalah anak dari class Hero
class Mage(Hero):
    def __init__(self, name, hp, attack_power, mana):
        # Memanggil constructor milik Parent (Hero)
        super().__init__(name, hp, attack_power)
        self.mana = mana

    def info(self):
        print(f"{self.name} [Mage] | HP: {self.hp} | Mana: {self.mana}")

    # Mage punya skill khusus
    def skill_fireball(self, lawan):
        if self.mana >= 20:
            print(f"{self.name} menggunakan Fireball ke {lawan.name}!")
            self.mana -= 20
            lawan.diserang(self.attack_power * 2)  # Damage 2x lipat
        else:
            print(f"{self.name} gagal skill! Mana tidak cukup.")


# -- Main Program Baru --
print("\n--- Update Class Hero ---")

eudora = Mage("Eudora", 80, 30, 10)
balmond = Hero("Balmond", 200, 10)

eudora.info()
eudora.serang(balmond)          # Serangan biasa (warisan dari Hero)
eudora.skill_fireball(balmond)  # Skill khusus Mage


--- Update Class Hero ---
Eudora [Mage] | HP: 80 | Mana: 10
Eudora menyerang Balmond!
Balmond terkena damage 30. Sisa HP: 170
Eudora gagal skill! Mana tidak cukup.


Penyebabnya, super init dihapus sehingga constructor Hero tidak dijalankan. Akibatnya, atribut name, hp, dan attack_power tidak dibuat pada objek Mage.

Fungsi super digunakan untuk memanggil constructor atau method dari class Induk, sehingga data dari class Anak dapat diproses, dan atribut milik class Induk dapat digunakan oleh class Anak.

### **LATIHAN SOAL 4**

In [4]:
class Hero:
    def __init__(self, nama, hp_awal):
        self.nama = nama

        # Enkapsulasi: HP bersifat private
        self.__hp = hp_awal

    # GETTER: Cara resmi melihat HP
    def get_hp(self):
        return self.__hp

    # SETTER: Cara resmi mengubah HP (dengan validasi)
    def set_hp(self, nilai_baru):
        if nilai_baru < 0:
            self.__hp = 0  # HP tidak boleh negatif
        elif nilai_baru > 1000:
            print("Cheat terdeteksi! HP dimaksimalkan ke 1000 saja.")
            self.__hp = 1000
        else:
            self.__hp = nilai_baru

    def diserang(self, damage):
        # Menggunakan getter dan setter agar data tetap aman
        sisa_hp = self.get_hp() - damage
        self.set_hp(sisa_hp)
        print(f"{self.nama} terkena damage {damage}. Sisa HP: {self.get_hp()}")


# -- Uji Coba --
hero1 = Hero("Layla", 100)

# hero1.__hp = 9999  # Gagal mengubah HP asli
# print(hero1.__hp)  # Error karena __hp bersifat private

hero1.set_hp(-50)  # Mencoba mengisi HP dengan nilai negatif
print(hero1.get_hp())  # Output: 0

0


Error Mage object has no attribute 'name' muncul karena pada Hero atributnya bernama self.nama, bukan self.name. Jadi eudora.info mencari self.name yang tidak ada.

Fungsi super digunakan untuk memanggil constructor class induk Hero, sehingga data seperti nama dan HP dari class Mage dapat diinisialisasi oleh Hero.



Hal ini membuat data menjadi tidak valid karena HP seharusnya tidak boleh bernilai negatif. Oleh karena itu, method setter sangat penting karena berfungsi memvalidasi data sebelum disimpan. Juga menghindari Cheating dengan darah 0 / -100 oleh pemain

### **LATIHAN SOAL 5**

In [ ]:
from abc import ABC, abstractmethod

# 1. Interface / Abstract Class
# Ini adalah KONTRAK. Semua turunan wajib memiliki method di bawah ini.
class GameUnit(ABC):

    @abstractmethod
    def serang(self, target):
        pass

    @abstractmethod
    def info(self):
        pass


# 2. Implementasi pada Class Hero
class Hero(GameUnit):

    def __init__(self, nama):
        self.nama = nama

    # Implementasi method serang
    def serang(self, target):
        print(f"Hero {self.nama} menebas {target}!")

    # Implementasi method info
    def info(self):
        print(f"Saya adalah Hero: {self.nama}")


# 3. Implementasi pada Class Monster
class Monster(GameUnit):

    def __init__(self, jenis):
        self.jenis = jenis

    # Implementasi method serang versi Monster
    def serang(self, target):
        print(f"Monster {self.jenis} menggigit {target}!")

    # Implementasi method info
    def info(self):
        print(f"Saya adalah Monster: {self.jenis}")


# -- Uji Coba --

# unit = GameUnit()  # ERROR! Abstract class tidak bisa dijadikan objek.

h = Hero("Alucard")
m = Monster("Serigala")

h.info()
m.info()

# Contoh penggunaan method serang
h.serang("Serigala")
m.serang("Alucard")

Saya adalah Hero: Alucard
Saya adalah Monster: Serigala
Hero Alucard menebas Serigala!
Monster Serigala menggigit Alucard!


Jika kita lupa membuat method yang sudah dijanjikan di interface (abstract class), maka class turunan tidak bisa digunakan untuk membuat objek. Hal

Class GameUnit tidak boleh dibuat menjadi objek karena merupakan abstract class. Abstract class hanya berfungsi sebagai cetakan (template) atau kontrak bagi class turunannya. Class ini belum memiliki isi dari method serang() dan info(), sehingga belum dapat digunakan sebagai objek nyata.

### **LATIHAN SOAL 6**

In [ ]:
# Parent Class
class Hero:
    def __init__(self, nama):
        self.nama = nama

    def serang(self):
        print("Hero menyerang dengan tangan kosong.")


# Child Class 1
class Mage(Hero):
    def serang(self):
        print(f"{self.nama} (Mage) menembakkan Bola Api! Boom!")


# Child Class 2
class Archer(Hero):
    def serang(self):
        print(f"{self.nama} (Archer) memanah dari jauh! Jleb!")


# Child Class 3
class Fighter(Hero):
    def serang(self):
        print(f"{self.nama} (Fighter) memukul dengan pedang! Slash!")


# -- Penerapan Polymorphism --

# Kita punya daftar hero campuran
pasukan = [
    Mage("Eudora"),
    Archer("Miya"),
    Fighter("Zilong"),
    Mage("Gord")
]

print("--- PERANG DIMULAI ---")

# Satu perintah loop, tetapi respon berbeda-beda (Polymorphism)
for pahlawan in pasukan:
    pahlawan.serang()

--- PERANG DIMULAI ---
Eudora (Mage) menembakkan Bola Api! Boom!
Miya (Archer) memanah dari jauh! Jleb!
Zilong (Fighter) memukul dengan pedang! Slash!
Gord (Mage) menembakkan Bola Api! Boom!


Keuntungan polimorfisme adalah programmer dapat menambahkan karakter baru tanpa harus mengubah kode yang sudah ada, terutama bagian looping. Selama class baru mengikuti struktur yang sama dan memiliki method serang()

Semua objek dalam pasukan diharapkan memiliki method bernama serang(). Namun, Archer sudah tidak memiliki method tersebut karena namanya diganti menjadi tembak_panah().

Dalam konsep polimorfisme, nama method antara Parent Class dan Child Class harus sama agar satu perintah dapat digunakan untuk berbagai jenis objek.

### **PROYEK INTEGRASI**

In [ ]:
from abc import ABC, abstractmethod


# ============================================================
# ABSTRACT CLASS
# ============================================================

class KamarHotel(ABC):

    def __init__(self, nama_kamar, stok, harga_dasar):
        self.nama_kamar = nama_kamar
        self.__stok = stok
        self.__harga_dasar = harga_dasar

    # Getter untuk melihat stok
    def get_stok(self):
        return self.__stok

    # Getter untuk melihat harga dasar
    def get_harga_dasar(self):
        return self.__harga_dasar

    # Method untuk menambah stok dengan validasi
    def tambah_stok(self, jumlah):
        stok_baru = self.__stok + jumlah

        if stok_baru < 0:
            print(
                f"Gagal update stok {self.nama_kamar}! "
                f"Stok tidak boleh negatif ({stok_baru})."
            )
        else:
            self.__stok = stok_baru
            print(
                f"Berhasil menambahkan stok {self.nama_kamar}: "
                f"{jumlah} unit."
            )

    # Abstract method
    @abstractmethod
    def tampilkan_detail(self):
        pass

    # Abstract method
    @abstractmethod
    def hitung_harga_total(self, jumlah_malam):
        pass


# ============================================================
# CLASS KAMAR DELUXE
# ============================================================

class KamarDeluxe(KamarHotel):

    def __init__(self, nama_kamar, stok, harga_dasar, fasilitas):
        super().__init__(nama_kamar, stok, harga_dasar)
        self.fasilitas = fasilitas

    # Override method tampilkan_detail()
    def tampilkan_detail(self):
        print(
            f"[DELUXE] {self.nama_kamar} | "
            f"Fasilitas: {self.fasilitas}"
        )

    # Override method hitung_harga_total()
    def hitung_harga_total(self, jumlah_malam):
        harga_dasar = self.get_harga_dasar()
        pajak = harga_dasar * 0.10
        harga_per_malam = harga_dasar + pajak
        subtotal = harga_per_malam * jumlah_malam

        print(
            f"   Harga Dasar/Malam: Rp {harga_dasar:,.0f} | "
            f"Pajak(10%): Rp {pajak:,.0f}"
        )
        print(
            f"   Menginap: {jumlah_malam} malam | "
            f"Subtotal: Rp {subtotal:,.0f}"
        )

        return subtotal


# ============================================================
# CLASS KAMAR STANDARD
# ============================================================

class KamarStandard(KamarHotel):

    def __init__(self, nama_kamar, stok, harga_dasar, kapasitas):
        super().__init__(nama_kamar, stok, harga_dasar)
        self.kapasitas = kapasitas

    # Override method tampilkan_detail()
    def tampilkan_detail(self):
        print(
            f"[STANDARD] {self.nama_kamar} | "
            f"Kapasitas: {self.kapasitas}"
        )

    # Override method hitung_harga_total()
    def hitung_harga_total(self, jumlah_malam):
        harga_dasar = self.get_harga_dasar()
        pajak = harga_dasar * 0.05
        harga_per_malam = harga_dasar + pajak
        subtotal = harga_per_malam * jumlah_malam

        print(
            f"   Harga Dasar/Malam: Rp {harga_dasar:,.0f} | "
            f"Pajak(5%): Rp {pajak:,.0f}"
        )
        print(
            f"   Menginap: {jumlah_malam} malam | "
            f"Subtotal: Rp {subtotal:,.0f}"
        )

        return subtotal


# ============================================================
# FUNGSI PROSES TRANSAKSI
# ============================================================

def proses_transaksi(daftar_pesanan):
    total_tagihan = 0

    print("\n--- STRUK PEMESANAN ---")

    for nomor, pesanan in enumerate(daftar_pesanan, start=1):
        kamar = pesanan["kamar"]
        jumlah_malam = pesanan["malam"]

        print(f"\n{nomor}. ", end="")
        kamar.tampilkan_detail()

        subtotal = kamar.hitung_harga_total(jumlah_malam)

        total_tagihan += subtotal

    print("\n--------------------------------")
    print(f"TOTAL TAGIHAN: Rp {total_tagihan:,.0f}")


# ============================================================
# MAIN PROGRAM
# ============================================================

print("--- SETUP DATA KAMAR ---")

# Admin membuat data kamar
deluxe = KamarDeluxe(
    "Kamar Deluxe Sea View",
    0,
    1_500_000,
    "Private Pool"
)

standard = KamarStandard(
    "Kamar Standard Superior",
    0,
    500_000,
    "2 Orang"
)


# Admin menambahkan stok
deluxe.tambah_stok(10)

# Mencoba memasukkan stok negatif
standard.tambah_stok(-5)

# Menambahkan stok yang benar
standard.tambah_stok(20)


# ============================================================
# PEMESANAN
# ============================================================

daftar_pesanan = [
    {
        "kamar": deluxe,
        "malam": 2
    },
    {
        "kamar": standard,
        "malam": 1
    }
]


# Memproses transaksi
proses_transaksi(daftar_pesanan)

--- SETUP DATA KAMAR ---
Berhasil menambahkan stok Kamar Deluxe Sea View: 10 unit.
Gagal update stok Kamar Standard Superior! Stok tidak boleh negatif (-5).
Berhasil menambahkan stok Kamar Standard Superior: 20 unit.

--- STRUK PEMESANAN ---

1. [DELUXE] Kamar Deluxe Sea View | Fasilitas: Private Pool
   Harga Dasar/Malam: Rp 1,500,000 | Pajak(10%): Rp 150,000
   Menginap: 2 malam | Subtotal: Rp 3,300,000

2. [STANDARD] Kamar Standard Superior | Kapasitas: 2 Orang
   Harga Dasar/Malam: Rp 500,000 | Pajak(5%): Rp 25,000
   Menginap: 1 malam | Subtotal: Rp 525,000

--------------------------------
TOTAL TAGIHAN: Rp 3,825,000
